constraint-solver/src/solver.rs

```diff
@@ -96,8 +110,187 @@ pub enum Error {
 /// An assignment of a variable.
 pub type VariableAssignment<T, V> = (V, GroupedExpression<T, V>);
 
+/// We introduce new variables.
+/// This enum avoids clashes with the original variables.
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub enum Variable<V> {
+    /// A regular variable that also exists in the original system.
+    Original(V),
+    /// A new boolean-constrained variable that was introduced by the solver.
+    Boolean(usize),
+}
+
+impl<V> From<V> for Variable<V> {
+    /// Converts a regular variable to a `Variable`.
+    fn from(v: V) -> Self {
+        Variable::Original(v)
+    }
+}
+
+impl<V: Clone> Variable<V> {
+    pub fn try_to_original(&self) -> Option<V> {
+        match self {
+            Variable::Original(v) => Some(v.clone()),
+            Variable::Boolean(_) => None,
+        }
+    }
+}
+
+impl<V: Display> Display for Variable<V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Variable::Original(v) => write!(f, "{v}"),
+            Variable::Boolean(i) => write!(f, "boolean_{i}"),
+        }
+    }
+}
+
+struct BooleanVarDispenser {
+    next_boolean_id: usize,
+}
+
+impl BooleanVarDispenser {
+    fn new() -> Self {
+        BooleanVarDispenser { next_boolean_id: 0 }
+    }
+
+    fn next_var<V>(&mut self) -> Variable<V> {
+        let id = self.next_boolean_id;
+        self.next_boolean_id += 1;
+        Variable::Boolean(id)
+    }
+
+    fn latest_var<V>(&self) -> Option<Variable<V>> {
+        (self.next_boolean_id > 0).then(|| Variable::Boolean(self.next_boolean_id - 1))
+    }
+}
+
+struct BooleanExtractedSolver<T, V, S> {
+    solver: S,
+    boolean_var_dispenser: BooleanVarDispenser,
+    _phantom: std::marker::PhantomData<(T, V)>,
+}
+
+impl<T, V, S> BooleanExtractedSolver<T, V, S>
+where
+    T: RuntimeConstant + VarTransformable<V, Variable<V>>,
+    T::Transformed: RuntimeConstant<FieldType = T::FieldType>,
+    V: Clone + Eq,
+    S: Solver<T::Transformed, Variable<V>>,
+{
+    fn new(solver: S) -> Self {
+        Self {
+            solver,
+            boolean_var_dispenser: BooleanVarDispenser::new(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T, V, S> RangeConstraintProvider<T::FieldType, V> for BooleanExtractedSolver<T, V, S>
```

Comment:
This is not used! There is a trait bound `Solver<T: RuntimeConstant, V>: RangeConstraintProvider<T::FieldType, V>`, but I tried removing it and it still compiles.

Reply:
Not sure what you mean? No one uses the `RangeConstraintProvider<T::FieldType, V>` trait on the solver, so the trait bound and implementation can be removed.